libbpf-tools/schedsnoop.c

```diff
@@ -0,0 +1,309 @@
+// SPDX-Licence-Identifier: GPL-2.0
+// Copyright (c) 2020 Alibaba Cloud
+//
+// Based on schedsnoop by Michael Wang
+// Created by Wenhao Qu
+//
+// Maintainers:
+// Michael Wang <yun.wang@linux.alibaba.com>
+// Wenhao Qu <quxi.qwh@alibaba-inc.com>
+#include <argp.h>
+#include <stdio.h>
+#include <signal.h>
+#include <sched.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <locale.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+#include <sys/wait.h>
+#include <dirent.h>
+#include <bpf/bpf.h>
+#include "errno_helpers.h"
+#include "trace_helpers.h"
+#include "syscall_helpers.h"
+#include "schedsnoop.h"
+#include "schedsnoop.skel.h"
+
+#define NS_IN_SEC		1000000000LLU
+#define NS_IN_MS		1000000LLU
+#define NS_IN_US		1000LLU
+#define PERF_BUFFER_PAGES	64
+
+const char *argp_program_version = "schedsnoop 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+static const char argp_program_doc[] =
+"Trace the related schedule events of a specified task.\n"
+"\n"
+"USAGE: schedsnoop -p PID [-s]\n"
+"\n"
+"EXAMPLES:\n"
+"    schedsnoop -p 49870       	# trace pid 49870\n"
+"    schedsnoop -p 49870 -s    	# trace pid 49870 and system call\n";
+
+static const struct argp_option opts[] = {
+	{ "pid", 'p', "PID", 0, "Process PID to trace"},
+	{ "syscall", 's', NULL, 0, "Trace SYSCALL Info"},
+	{},
+};
+
+static struct env {
+	int nr_cpus;
+	int trace_syscall;
+	int target;
+} env;
+
+int ti_map_fd;
+int si_map_fd;
+int start_map_fd;
+int end_map_fd;
+bool exiting = false;
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int pid;
+	switch (key) {
+	case 'p':
+		if(arg == NULL) {
+			fprintf(stderr, "PID is required\n");
+			argp_usage(state);
+		} else {
+			pid = strtol(arg, NULL, 10);
+			if (pid <= 0) {
+				fprintf(stderr, "Invalid PID: %s\n", arg);
+				argp_usage(state);
+			}
+			env.target = pid;
+		}
+		break;
+	case 's':
+		env.trace_syscall = 1;
+		break;
+	case ARGP_KEY_END:
+		if (!env.target) {
+			fprintf(stderr, "No target PID\n");
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static inline int time_to_str(u64 ns, char *buf, size_t len)
+{
+
+	if (ns > 10 * NS_IN_SEC)
+		snprintf(buf, len, "%llus", ns / NS_IN_SEC);
+	else if (ns > 10 * NS_IN_MS)
+		snprintf(buf, len, "%llums", ns / NS_IN_MS);
+	else if (ns > 10 * NS_IN_US)
+		snprintf(buf, len, "%lluus", ns / NS_IN_US);
+	else
+		snprintf(buf, len, "%lluns", ns);
+
+	return 0;
+}
+
+static inline void pr_ti(struct trace_info *ti, char *opt, char *delay)
+{
+	char buf[27];
+	snprintf(buf, sizeof(buf), "%lluus", ti->ts / NS_IN_US);

```

Comment:
so you are still printing absolute timestamp and then optionally prepend relative timestamp? why both? why not just using relative all the time?

Reply:
the relative timestamp is not the time different between two adjacent events. It is used to show how long our target task has been executed, how long another task has preempted the target task. It is optional because some events don't have a duration.

the absolute timestamp is not so important here, but we believe it is necessary to keep it. the absolute timestamp may be helpful for developer, especially kernel developer, to trace the system and locate the problem.

Reply:
Maybe I can add a debug options in argp and only output raw timestamp when dubug is on.

Reply:
Yeah, command line option to switch between relative vs absolute timestamps make sense. Relative is more user-friendly by default, IMO.